homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,489 @@
+
+            if self.data[DATA_CHARGING]:
+                intervals.append(charging_interval)
+                _LOGGER.debug("Could use charging interval=%s",
+                              charging_interval)
+
+            if self.data[DATA_CLIMATE]:
+                intervals.append(climate_interval)
+                _LOGGER.debug("Could use climate interval=%s",
+                              climate_interval)
+
+            interval = min(intervals)
+            _LOGGER.debug("Resulting interval=%s", interval)
+
+        return utcnow() + interval
+
+    async def async_refresh_data(self, now):
+        """Refresh the leaf data and update the datastore."""
+        from pycarwings2 import CarwingsError
+
+        if self.request_in_progress:
+            _LOGGER.debug("Refresh currently in progress for %s",
+                          self.leaf.nickname)
+            return
+
+        _LOGGER.debug("Updating Nissan Leaf Data")
+
+        self.last_check = datetime.today()
+        self.request_in_progress = True
+
+        (battery_response, server_response) = await self.async_get_battery()
+
+        if battery_response is not None:
+            _LOGGER.debug("Battery Response: ")
+            _LOGGER.debug(battery_response.__dict__)
+
+            if battery_response.answer['status'] == 200:
+                if int(battery_response.battery_capacity) > 100:
+                    self.data[DATA_BATTERY] = (
+                        battery_response.battery_percent * 0.05
+                    )
+                else:
+                    self.data[DATA_BATTERY] = battery_response.battery_percent
+
+                self.data[DATA_CHARGING] = battery_response.is_charging
+                self.data[DATA_PLUGGED_IN] = battery_response.is_connected
+                self.data[DATA_RANGE_AC] = (
+                    battery_response.cruising_range_ac_on_km
+                )
+                self.data[DATA_RANGE_AC_OFF] = (
+                    battery_response.cruising_range_ac_off_km
+                )
+                self.signal_components()
+                self.last_battery_response = utcnow()
+
+        if server_response is not None:
+            _LOGGER.debug("Server Response: ")
+            _LOGGER.debug(server_response.__dict__)
+
+            if server_response.answer['status'] == 200:
+                if server_response.state_of_charge is not None:
+                    self.data[DATA_BATTERY] = int(
+                        server_response.state_of_charge
+                    )
+                else:
+                    self.data[DATA_BATTERY] = server_response.battery_percent
+
+                self.data[DATA_RANGE_AC] = (
+                    server_response.cruising_range_ac_on_km
+                )
+                self.data[DATA_RANGE_AC_OFF] = (
+                    server_response.cruising_range_ac_off_km
+                )
+                self.signal_components()
+                self.last_battery_response = utcnow()
```

Comment:
This code is tase same as the previous one, probably can be refactored ?

Reply:
Should be fixed by commit https://github.com/home-assistant/home-assistant/pull/19786/commits/33fcd3f60e2cda2ba7b25aeb8d0b39c892ea008a

Reply:
I was referring to all the code that checks battery_response.

Reply:
Yes, sorry. Thanks. Since the nissan servers have changed the way they respond I've simplified the code and only consider them returning via one route, so we only have to handle it one type of response in `async_refresh_data` and `async_get_battery`.